MetadataExtractor/Formats/Jpeg/HuffmanTablesDirectory.cs

```diff
@@ -0,0 +1,323 @@
+        /// and some hardware decoders.
+        /// <para/>
+        /// Even though the JPEG standard doesn't define these as "standard tables"
+        /// and requires a decoder to be able to read any valid Huffman tables, some
+        /// are in reality limited decoding images using these "typical" tables.
+        /// Standards like DCF(Design rule for Camera File system) and DLNA(Digital
+        /// Living Network Alliance) actually requires any compliant JPEG to use only
+        /// the "typical" Huffman tables.
+        /// <para/>
+        /// This is also related to the term "optimized" JPEG. An "optimized" JPEG is
+        /// a JPEG that doesn't use the "typical" Huffman tables.
+        /// </remarks>
+        public bool IsTypical()
+        {
+            if (Tables.Count == 0)
+            {
+                return false;
+            }
+            foreach (HuffmanTable table in Tables)
+            {
+                if (!table.IsTypical())
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /// <remarks>The opposite of IsTypical().</remarks>
+        /// <returns>
+        /// Whether or not the tables in this HuffmanTablesDirectory
+        /// are "optimized" - which means that at least one of them aren't
+        /// one of the "typical" Huffman tables.
+        /// </returns>
+        public bool IsOptimized()
+        {
+            return !IsTypical();
+        }
+
+    }
+
+    /// <summary>An instance of this class holds a JPEG Huffman table.</summary>
+    public class HuffmanTable
+    {
+        private byte[] _lengthBytes;
+        private byte[] _valueBytes;
+
+        public HuffmanTable([NotNull] HuffmanTableClass tableClass, int tableDestinationId, [NotNull] byte[] lengthBytes, [NotNull] byte[] valueBytes)
+        {
+            _lengthBytes = lengthBytes ?? throw new ArgumentNullException("lengthBytes cannot be null.");
+            _valueBytes = valueBytes ?? throw new ArgumentNullException("valueBytes cannot be null.");
+
+            TableClass = tableClass;
+            TableDestinationId = tableDestinationId;
+            TableLength = _valueBytes.Length + 17;
+        }
+
+        /// <returns>The table length in bytes.</returns>
+        public int TableLength { get; private set; }
+
+        /// <returns>The HuffmanTableClass of this table.</returns>
+        public HuffmanTableClass TableClass { get; private set; }
+
+        /// <returns>The destination identifier for this table.</returns>
+        public int TableDestinationId { get; private set; }
+
+        /// <returns>A byte array with the L values for this table.</returns>
+        [NotNull]
+        public byte[] LengthBytes
+        {
+            get
+            {
+                byte[] result = new byte[_lengthBytes.Length];
+                Array.Copy(_lengthBytes, 0, result, 0, _lengthBytes.Length);
+                return result;
```

Comment:
Can this just be `_lengthBytes.ToArray()`?

Reply:
I'm not sure what Nadahar intended for LengthBytes and ValueBytes. Given that they do a copy, maybe it was to give access to the array contents without allowing modifications.

I'll stay with the spirit of the original Java and return a copy, but will make it ToArray as you suggest.

Reply:
I think you're right, that it was a defensive copy to prevent external code from modifying the contents.

If/when we update the versions of .NET we support we can change these to `ReadOnlySpan<char>`, though unfortunately that's likely a way off (in a major release).